File: javav2/example_code/sns/src/main/java/com/example/sns/Unsubscribe.java
//snippet-sourcedescription:[Unsubscribe.java demonstrates how to remove an Amazon Simple Notification Service (Amazon SNS) subscription.]
//snippet-keyword:[AWS SDK for Java v2]
//snippet-keyword:[Code Sample]
//snippet-keyword:[Amazon Simple Notification Service]
//snippet-sourcetype:[full-example]
//snippet-sourcedate:[05/19/2022]

/*
   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
   SPDX-License-Identifier: Apache-2.0
*/

package com.example.sns;

//snippet-start:[sns.java2.Unsubscribe.import]
import software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.sns.SnsClient;
import software.amazon.awssdk.services.sns.model.SnsException;
import software.amazon.awssdk.services.sns.model.UnsubscribeRequest;
import software.amazon.awssdk.services.sns.model.UnsubscribeResponse;
//snippet-end:[sns.java2.Unsubscribe.import]

/**
 * Before running this Java V2 code example, set up your development environment, including your credentials.
 *
 * For more information, see the following documentation topic:
 *
 * https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/get-started.html
 */
public class Unsubscribe {

    public static void main(String[] args) {

        final String usage = "\n" +
                "Usage: " +
                "   <subscriptionArn>\n\n" +
                "Where:\n" +
                "   subscriptionArn - The ARN of the subscription to delete.\n\n";

        if (args.length < 1) {
            System.out.println(usage);
            System.exit(1);
        }

        String subscriptionArn = args[0];
        SnsClient snsClient = SnsClient.builder()
                .region(Region.US_EAST_1)
                .credentialsProvider(ProfileCredentialsProvider.create())
                .build();

        unSub(snsClient, subscriptionArn);
        snsClient.close();
    }

    //snippet-start:[sns.java2.Unsubscribe.main]
    public static void unSub(SnsClient snsClient, String subscriptionArn) {

        try {
            UnsubscribeRequest request = UnsubscribeRequest.builder()
                .subscriptionArn(subscriptionArn)
                .build();

            UnsubscribeResponse result = snsClient.unsubscribe(request);

            System.out.println("\n\nStatus was " + result.sdkHttpResponse().statusCode()
                + "\n\nSubscription was removed for " + request.subscriptionArn());

        } catch (SnsException e) {
            System.err.println(e.awsErrorDetails().errorMessage());
            System.exit(1);
        }
    }
    //snippet-end:[sns.java2.Unsubscribe.main]
}

File: kotlin/services/sns/src/main/kotlin/com/kotlin/sns/Unsubscribe.kt
// snippet-sourcedescription:[Unsubscribe.kt demonstrates how to remove an Amazon Simple Notification Service (Amazon SNS) subscription.]
// snippet-keyword:[AWS SDK for Kotlin]
// snippet-keyword:[Amazon Simple Notification Service]

/*
   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
   SPDX-License-Identifier: Apache-2.0
*/

package com.kotlin.sns

// snippet-start:[sns.kotlin.Unsubscribe.import]
import aws.sdk.kotlin.services.sns.SnsClient
import aws.sdk.kotlin.services.sns.model.UnsubscribeRequest
import kotlin.system.exitProcess
// snippet-end:[sns.kotlin.Unsubscribe.import]

/**
Before running this Kotlin code example, set up your development environment,
including your credentials.

For more information, see the following documentation topic:
https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/setup.html
 */
suspend fun main(args: Array<String>) {

    val usage = """
        Usage: 
            <subscriptionArn>

        Where:
            subscriptionArn - The ARN of the subscription.
        """

    if (args.size != 1) {
        println(usage)
        exitProcess(0)
    }

    val subArn = args[0]
    unSub(subArn)
}

// snippet-start:[sns.kotlin.Unsubscribe.main]
suspend fun unSub(subscriptionArnVal: String) {

    val request = UnsubscribeRequest {
        subscriptionArn = subscriptionArnVal
    }

    SnsClient { region = "us-east-1" }.use { snsClient ->
        snsClient.unsubscribe(request)
        println("Subscription was removed for ${request.subscriptionArn}")
    }
}
// snippet-end:[sns.kotlin.Unsubscribe.main]
